tls/s2n_record_read_stream.c

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "crypto/s2n_sequence.h"
+#include "crypto/s2n_cipher.h"
+#include "crypto/s2n_hmac.h"
+
+#include "error/s2n_errno.h"
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_crypto.h"
+#include "tls/s2n_record_read.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_blob.h"
+
+
+int s2n_record_parse_stream(const struct s2n_cipher_suite *cipher_suite,
+			    struct s2n_connection *conn,
+			    uint8_t content_type,
+			    uint16_t encrypted_length,
+			    uint8_t *implicit_iv,
+			    struct s2n_hmac_state *mac,
+			    uint8_t *sequence_number,
+			    struct s2n_session_key *session_key)
+{
+    /* Add the header to the HMAC */
+    uint8_t *header = s2n_stuffer_raw_read(&conn->header_in, S2N_TLS_RECORD_HEADER_LENGTH);
+    notnull_check(header);
+
+    struct s2n_blob en = {.size = encrypted_length, .data = s2n_stuffer_raw_read(&conn->in, encrypted_length)};
+    notnull_check(en.data);
+
+    uint16_t payload_length = encrypted_length;
+    uint8_t mac_digest_size;
+    GUARD(s2n_hmac_digest_size(mac->alg, &mac_digest_size));
+
+    gte_check(payload_length, mac_digest_size);
+    payload_length -= mac_digest_size;
+
+    /* Decrypt stuff! */
+    GUARD(cipher_suite->record_alg->cipher->io.stream.decrypt(session_key, &en, &en));
+
+    /* Update the MAC */
```

Comment:
This identical code also appears for CBC mode.  Would people prefer if I made this its own function, or would that be less clear?

Reply:
meh, i don't feel strongly about this.  i think it is fine to have some small amounts of duplicated code, or you get too abstract :)